### PR TITLE
Adding parametric model and training scripts

### PR DESCRIPTION
**Pull request by jsschreck**

### CODE REVIEW

**Review by ggantos — APPROVED** [review]
Looks good - there are a few small things that I might add in my next PR like doc strings for load_preprocessor vs load_preprocessing but these seem minor.

File: evml/metrics.py [hunk 1]
```diff
@@ -0,0 +1,624 @@
...
+    pit = np.zeros_like(pred_sorted)
+    for i, x in enumerate(pred_sorted):
+        j = np.searchsorted(true_sorted, x, side="right")
+        pit[i] = j / len(true_sorted)
+
+    return pit
+
+
+def pit_histogram(
+    df, pred_dist, output_cols, num_bins=50, legend_cols=None, save_location=None
```
Comment (djgagne) [inline evml/metrics.py]: What is the shape of pred_dist? Does it consist of ensemble members or distribution parameters?

File: evml/metrics.py [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+        axs = [axs]
+
+    if legend_cols is None:
+        legend_cols = output_cols
+
+    true_dist = df[output_cols].values
+
+    for k in range(true_dist.shape[-1]):
+        td = true_dist[:, k]
+        pd = pred_dist[:, k]
```
Comment (djgagne) [inline evml/metrics.py]: For a given output prediction, shouldn't there be more than one ensemble member? The way the slicing is done here makes it look like there is one prediction per example.

File: evml/metrics.py [hunk 3]
```diff
@@ -0,0 +1,624 @@
...
+
+    Returns
+    -------
+    pit : array-like
+        A 1D array of PIT values.
+    """
+    # Sort the true and predicted distributions
+    true_sorted = np.sort(true_dist)
+    pred_sorted = np.sort(pred_dist)
+
```
Comment (djgagne) [inline evml/metrics.py]: If these are 1D arrays, then sorting them would break the relationship between true and pred.

File: evml/metrics.py [hunk 4]
```diff
@@ -0,0 +1,624 @@
...
+    # Compute the PIT values
+    pit = np.zeros_like(pred_sorted)
+    for i, x in enumerate(pred_sorted):
+        j = np.searchsorted(true_sorted, x, side="right")
```
Comment (djgagne) [inline evml/metrics.py]: The PIT verification metric is finding the quantile of the truth in the predicted distribution. This code is finding the quantile of the predicted value in the true distribution.

**Review by djgagne — CHANGES_REQUESTED** [review]
The PIT code is not correct for the purposes of verification. See my other comments on fixing them.

File: evml/keras/monte_carlo.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+    # Calculating mean across multiple MCD forward passes
+    mean_mu = np.mean(dropout_mu, axis=0)  # shape (n_samples, n_classes)
+    #mean_aleatoric = np.mean(dropout_aleatoric, axis=0)  # shape (n_samples, n_classes)
+
+    # Calculating variance across multiple MCD forward passes
+    variance_mu = np.var(dropout_mu, axis=0)  # shape (n_samples, n_classes)
+    #variance_aleatoric = np.var(dropout_aleatoric, axis=0)  # shape (n_samples, n_classes)
+
+    return mean_mu, variance_mu
```
Comment (djgagne) [inline evml/keras/monte_carlo.py]: Not sure if calculations are correct. See supplementary material for this [paper](https://agupubs.onlinelibrary.wiley.com/doi/full/10.1029/2022GL098566) to get a formula for aleatoric and epistemic from a parametric neural network with Monte Carlo dropout.

File: evml/keras/models.py [hunk 6]
```diff
@@ -71,70 +95,218 @@ def build_neural_network(self, inputs, outputs):
...
+
         if y_scaler:
-            mu = y_scaler.inverse_transform(mu.reshape((mu.shape[0], -1))).squeeze()
+            mu = y_scaler.inverse_transform(mu)
+
+        for i in range(mu.shape[-1]):
+            aleatoric[:, i] *= self.training_var[i]
+            epistemic[:, i] *= self.training_var[i]
+
+        return mu, np.sqrt(aleatoric), np.sqrt(epistemic)
```
Comment (djgagne) [inline evml/keras/models.py]: We should decide on convention for aleatoric and epistemic uncertainty and whether to keep in variance form or convert to standard deviation form as currently done here. Adding them together is more straightforward in variance form.

**Comment by djgagne** [comment]
I just added my pit.py and test_pit.py modules to your pull request.

File: evml/keras/models.py [hunk 7]
```diff
@@ -226,7 +231,17 @@ def calc_uncertainties(self, preds, y_scaler):
...
 
-        return mu, np.sqrt(aleatoric), np.sqrt(epistemic)
+        return mu, aleatoric, epistemic
+    
+    def predict_dist_params(self, x):
+        preds = self.model.predict(x, batch_size=self.batch_size)
+        mu, v, alpha, beta = np.split(preds, 4, axis=-1)
+        if mu.shape[-1] == 1:
+            mu = np.expand_dims(mu, 1)
+        if y_scaler:
```
Comment (djgagne) [inline evml/keras/models.py:429]: There is no y_scaler input to this function, resulting in the flake8 failure. Can you fix?

File: evml/metrics.py (hunk 4 above)
Comment (jsschreck) [reply evml/metrics.py]: Thanks for pushing the updates! They should be incorporated in the latest version.

File: evml/keras/monte_carlo.py (hunk 5 above)
Comment (jsschreck) [reply evml/keras/monte_carlo.py]: Yeah this was wrong -- I think we have it correctly implemented in the latest version.

File: evml/keras/models.py (hunk 6 above)
Comment (jsschreck) [reply evml/keras/models.py]: Yeah, agree, this has been incorporated starting with the model outputting the variances rather than std.

File: evml/keras/models.py (hunk 7 above)
Comment (jsschreck) [reply evml/keras/models.py:429]: Thanks; fixed!

**Review by djgagne — APPROVED** [review]

**Comment by djgagne** [comment]
Looks good to me. Merging.
